examples/jsm/loaders/MMDLoader.js

```diff
@@ -2064,4 +2076,114 @@ class CubicBezierInterpolation extends Interpolant {
 
 }
 
+class MMDToonMaterial extends ShaderMaterial {
+
+	constructor( parameters ) {
+
+		super();
+
+		this._matcapCombine = AddOperation;
+		this.emissiveIntensity = 1.0;
+		this.normalMapType = TangentSpaceNormalMap;
+
+		this.combine = MultiplyOperation;
+
+		this.wireframeLinecap = 'round';
+		this.wireframeLinejoin = 'round';
+
+		this.flatShading = false;
+
+		this.lights = true;
+
+		this.vertexShader = MMDToonShader.vertexShader;
+		this.fragmentShader = MMDToonShader.fragmentShader;
+
+		this.defines = Object.assign( {}, MMDToonShader.defines );
+		Object.defineProperties( this, {
+
+			matcapCombine: {
+
+				get: function () {
+
+					return this._matcapCombine;
+
+				},
+
+				set: function ( value ) {
+
+					this._matcapCombine = value;
+
+					switch ( value ) {
+
+						case MultiplyOperation:
+							this.defines.MATCAP_BLENDING_MULTIPLY = true;
+							delete this.defines.MATCAP_BLENDING_ADD;
+							break;
+
+						default:
+						case AddOperation:
+							this.defines.MATCAP_BLENDING_ADD = true;
+							delete this.defines.MATCAP_BLENDING_MULTIPLY;
+							break;
+
+					}
+
+				}
+
+			},
+
+		} );
+
+		this.uniforms = UniformsUtils.clone( MMDToonShader.uniforms );
+		for ( const uniformName in this.uniforms ) {
```

Comment:
I first thought of using getter/setter in class like

```javascript
class MMDToonMaterial {
  constructor(params) {
    ...
  }
  get foo() {
    return this.uniforms.foo.value;
  }
  set foo(value) {
    this.uniforms.foo.value = value;
  }
}
```

but your `Object.defineProperty` approach would be simpler.

You set up getter/setter for all uniforms but uniform includes some properties which don't need to be exposed like `directionalLightShadows`. So I think it would be better to set up getter/setter only for properties which should be exposed.

```javascript
const propertyNames = [
  ...
];
for ( const propName of propertyNames ) {
  ...
}
```

Reply:
Would it be better if I choose like this?

```js
const excludedProtertyNames = [
  directionalLightShadows,
  ...
];
for ( const propertyName of propertyNames ) {
  if ( !excludedProtertyNames.includes( propertyName ) ) {
    ...
  }
}
```

Make it default expose in case there will be more uniforms need to be exposed from MeshToon/Phong/MatcapMaterial in the future?

Reply:
I think it's pros and cons. Cons of the exclude approach is if new uniforms which shouldn't be exposed will be added to toon/phong/matcap uniforms they will be exposed.

Personally I prefer mine because it would be easier to notice the need for the update. But no strong opinion.

Reply:
I think yours should be better since it should be more understandable for someone wan't to modify this code.